jobs: job state change structured logging

kyle-a-wong opened the pull request:
Adds a new `status_change` job event that logs
on every status update.

Epic: [CRDB-37556](https://cockroachlabs.atlassian.net/browse/CRDB-37556)

Release note (general change): Job status changes
now log events to the OPS channel, indicating the
previous status and the new status.

xinhaoz commented on pkg/jobs/jobs_structured_log_test.go:
```diff
@@ -0,0 +1,100 @@
+	}
+	rts.job = j
+
+	rts.mu.e.ResumeStart = true
```
Is it safe to access this without locking the mutex? Same with the other access in `mu` on line 70/71.

xinhaoz commented on pkg/jobs/structured_log.go:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	if jobErr != nil {
+		// TODO(abarganier): properly redact
```
Can we provide a little more detail into what 'properly redact' entails here.

xinhaoz commented on pkg/util/log/logtestutils/structured_log_spy.go:
```diff
@@ -142,6 +142,10 @@ type StructuredLogSpy[T any] struct {
+
+		// lastChannelRead is a map of channel to int, representing the last read log
+		// line read when calling GetUnreadLogs.
+		lastChannelRead map[logpb.Channel]int
```
nit: Feel free to ignore just a personal preference, mostly. This naming reads like we're tracking the last channel read, not the channel to last time read. Maybe 'lastReadIdx' would be more clear.

xinhaoz replied:
This map also needs to be reset in the `Reset` fn, which clears all the logs.

xinhaoz commented on pkg/util/log/logtestutils/structured_log_spy.go:
```diff
@@ -237,6 +243,19 @@ func (s *StructuredLogSpy[T]) GetLogs(ch logpb.Channel) []T {
+// since the last time this method was called.
+func (s *StructuredLogSpy[T]) GetUnreadLogs(ch logpb.Channel) []T {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
```
Take the full lock since you're writing to lastChannelRead below.

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/util/log/logtestutils/structured_log_spy.go` line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1636699384:-O-CL_Ho2BjwnkOkvD6I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/04a08381af072e7a7f9ce8f129126f68d73a2fca/pkg/util/log/logtestutils/structured_log_spy.go#L148)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

This map also needs to be reset in the `Reset` fn, which clears all the logs.
</blockquote></details>

Done.
___
*[`pkg/util/log/logtestutils/structured_log_spy.go` line 250 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1636720589:-O-CLa0_BJMpeqxwrJ5v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/04a08381af072e7a7f9ce8f129126f68d73a2fca/pkg/util/log/logtestutils/structured_log_spy.go#L250)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Take the full lock since you're writing to lastChannelRead below.
</blockquote></details>

Done.
___
*[`pkg/jobs/jobs_structured_log_test.go` line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1636691272:-O-CLb1NBRTuoM_zPZ9J:by1nmd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/04a08381af072e7a7f9ce8f129126f68d73a2fca/pkg/jobs/jobs_structured_log_test.go#L65)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Is it safe to access this without locking the mutex? Same with the other access in `mu` on line 70/71.
</blockquote></details>

Honestly, I just copy pasted this from the `jobs_test` file where it did the same thing :sweat: . Update to lock.


<!-- Sent from Reviewable.io -->

dt commented on pkg/jobs/structured_log.go:
```diff
@@ -0,0 +1,78 @@
+
+type JobEvent struct {
+	JobID        jobspb.JobID `json:"job_id"`
+	EventType    JobEventType `json:"event_type"`
```
should we add some notion of notability or severity to these events to allow a reader of this log to more easily read for only notable events (like state changes or resuming after retries) vs all events?

I know we only have one event type now, but we might want to be writing it with some sort of level attached now so that when we start writing more events we have a level against which to set those?

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dt, @michae2, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/jobs/structured_log.go` line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1636694922:-O-D7FAc5e1GhxlV9DSO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/04a08381af072e7a7f9ce8f129126f68d73a2fca/pkg/jobs/structured_log.go#L60)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Can we provide a little more detail into what 'properly redact' entails here.
</blockquote></details>

Done.
___
*[`pkg/jobs/structured_log.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1638186023:-O-IAaj5Dnv-9YHBRYpZ:b-t8d8n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b9348cf82b182d6681db12b794dabd7bf80b8fb/pkg/jobs/structured_log.go#L32)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

should we add some notion of notability or severity to these events to allow a reader of this log to more easily read for only notable events (like state changes or resuming after retries) vs all events?

I know we only have one event type now, but we might want to be writing it with some sort of level attached now so that when we start writing more events we have a level against which to set those?
</blockquote></details>

Added an `EventLevel` field. I wasn't sure what exactly you had in mind in terms of what the different levels would mean, so I went with 1 being the most "severe" and 5 being the least. I also used nomenclature similar to log levels. Lmk if you have different ideas / opinions about this.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @kyle-a-wong, @michae2, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/jobs/structured_log.go` line 61 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O-IIwiB19WTRX5MoCgP:-O-IIwiB19WTRX5MoCgQ:b-s4qi8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc72b1016a8fe4cf4ab2ad4998f52eaf2bc97a29/pkg/jobs/structured_log.go#L61)):*
> ```Go
> const (
> 	JobEventLevelNone JobEventLevel = iota
> 	JobEventLevelFatal
> ```

Jobs can't fatal a node so can remove this one.


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dt, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/jobs/structured_log.go` line 61 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O-IIwiB19WTRX5MoCgP:-O-IUCBT8Scu9bzRibM6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc72b1016a8fe4cf4ab2ad4998f52eaf2bc97a29/pkg/jobs/structured_log.go#L61)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Jobs can't fatal a node so can remove this one.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

xinhaoz commented on pkg/jobs/jobs_structured_log_test.go:
```diff
@@ -0,0 +1,104 @@
+			b, _ := json.Marshal(jobEventsLog.EventDetails)
+			var stateChangeDetails jobs.JobStateChangeDetails
+			_ = json.Unmarshal(b, &stateChangeDetails)
+			if stateChangeDetails.NewStatus == expectedStatus {
```
Do we want to validate any other fields here, like EventDetails or EventLevel?

xinhaoz reviewed (approved):
Nice, lgtm!

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dt, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/jobs/jobs_structured_log_test.go` line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1640350889:-O-qcnET7EFRYYOAYHwN:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/92e71823b73ef34b679a0cfece1dc6fc356a04af/pkg/jobs/jobs_structured_log_test.go#L38)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

nit: should this be `jobSpy` / something more appropriate?
</blockquote></details>

done
___
*[`pkg/jobs/jobs_structured_log_test.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1640350956:-O-qas_A0mOg2_ZbJRsc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92e71823b73ef34b679a0cfece1dc6fc356a04af/pkg/jobs/jobs_structured_log_test.go#L97)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Why aren't we checking the error here?
</blockquote></details>

Done.
___
*[`pkg/jobs/jobs_structured_log_test.go` line 98 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1640352740:-O-qauNk5YZEXyJ2d3xI:b-58d16d) ([raw file](https://github.com/cockroachdb/cockroach/blob/92e71823b73ef34b679a0cfece1dc6fc356a04af/pkg/jobs/jobs_structured_log_test.go#L98)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Do we want to validate any other fields here, like EventDetails or EventLevel?
</blockquote></details>

added another validation for previous status


<!-- Sent from Reviewable.io -->

kyle-a-wong commented:
@msbutler In https://github.com/cockroachdb/cockroach/pull/124242 you mentioned the following:
> nit: no need to address this now, but sadly, each job status change may not go through the stepThroughStateMachine (e.g. pause), so we'll need to rethink where to call this.

Do you know where else we would need to call `maybeLogStateChangeStructured` ?

msbutler commented:
@kyle-a-wong so the jobs system will pause and cancel jobs via `servePauseAndCancelRequests` so in theory you could call `maybeLogStateChangeStructured` in there. However, I don't think you should call `maybeLogStateChangeStructured` within the txn, as the txn may retry. To get around this, you could create a separate method that creates and buffers  the relevant `JobEvent`s. On txn retry, you clear the buffer, and after the txn completes, you could log the batch of structured events. Does that make sense?  cc @dt

kyle-a-wong committed:
log,job: log status change on update via updater

Adds a new `status_change` job event that logs
on every status update.

Epic: CRDB-37556

Release note (general change): Job status changes
now log events to the OPS channel, indicating the
previous status and the new status.

kyle-a-wong commented:
Decided to go a different direction with this and use the existing protobuf based structured entries for these logs. 
Ive made the following changes to this PR:
1. removed new JobEvent struct
2. created a new status_change protobuf type
3. Change to do the logging for status update in `update.go` via a commit trigger
4. general changes to `structured_v2.go` to log structured entries and pass them to the log processor.

@msbutler @dt lmk what you think of where I'm calling `LogStateChangeStructured` and if you have any concerns about it

msbutler commented on pkg/jobs/update.go:
```diff
@@ -171,6 +171,17 @@ WHERE id = $1
 	if err := updateFn(u.txn, md, &ju); err != nil {
 		return err
 	}
+	if ju.md.Status != "" {
```
`update` is not just called during a status change, and further, md.Status will always be not nil. So, this current change will log structured events on many more events than a status change. I don't think this is intended?

msbutler replied:
I think your original call sites were fine, and calling this in a commit trigger is a nice way to avoid the retry problem!

kyle-a-wong replied:
One reason why i wanted to move it here is that the update function is guaranteed to be in a transaction which made it safe to use the commit trigger and we know that this function is always going to be invoked for status changes. 

From reading the code, `ju.md.Status` is only set on status updates, which i saw by checking usages of `JobUpdater.UpdateStatus`, but i guess its not good rely on this always being true in the future. What if we leave it here and change the condition to something like: 
`if ju.md.Status != "" && (status != ju.md.Status || (status == StatusRunning && ju.md.Status == StatusRunning)) {...}`

msbutler replied:
huh, I see. I'm coming around to this call site! While your argument does hold for all status changes that occur via `UpdateStatus()`, sadly we do not call `UpdateStatus(PAUSED)` or `UpdateStatus(REVERTING)` (for all paths) and instead manually change the status via this query. https://github.com/msbutler/cockroach/blob/butler-skip-auto-replan-stress/pkg/jobs/adopt.go#L530

I think if you added a call `maybeLogStateChangeStructured` around this query and added a little dosctring explaining the logic in the `update()` path, I'd feel good about this approach. @dt should also take a look at this.

kyle-a-wong replied:
Awesome, thanks for the feedback! I updated the conditional here, added a doc string, and added the log to `servePauseAndCancelRequests` as well.

msbutler replied:
Thanks for addressing this! I left a few more comments, but will not have the time to give this a full review before heading on PTO tomorrow. I'd ping @stevendanna or @dt for a final review.

msbutler commented on pkg/jobs/update.go:
```diff
@@ -171,6 +171,29 @@ WHERE id = $1
+	// required when a job starts because it may already have a "running" status.
+	//
+	if ju.md.Status != "" &&
+		(ju.md.Status != status || (ju.md.Status == StatusRunning && status == StatusRunning)) {
```
sorry, i should have said this earlier: I'm not following why `ju.md.Status == StatusRunning && status == StatusRunning` implies a status update?

kyle-a-wong replied:
This is so that we can capture a "job started"  status change. Im not sure the historical reasonings for why the status change looks like this, but a job can only "start" if the existing status of the job is `StatusPending` or `StatusRunning`, so this condition is expected when a job "starts" but is already in the running status.

https://github.com/cockroachdb/cockroach/blob/a6f53a802cafa063e906c18d85f57c13b2a98315/pkg/jobs/jobs.go#L253

msbutler commented on pkg/jobs/adopt.go:
```diff
@@ -511,7 +511,7 @@ const pauseAndCancelUpdate = `
     WHERE (status IN ('` + string(StatusPauseRequested) + `', '` + string(StatusCancelRequested) + `'))
       AND ((claim_session_id = $1) AND (claim_instance_id = $2))
-RETURNING id, status
+RETURNING id, status, job_type
```
i don't think we should change this query for this pr-- we can grab the job type from the in memory job object

msbutler replied:
@dt are you ok changing this query for this PR?

dhartunian reviewed (commented):
> Reviewed 3 of 11 files at r7, 2 of 6 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @kyle-a-wong, @msbutler, @stevendanna, and @xinhaoz)

___
*[`-- commits` line 15 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0kpT1a0xysWPUarBzn):*
nit: `OPS`
___
*[`pkg/jobs/structured_log.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1638186023:-O0sqQLg8_UWXK_PLvmi:b-vvj5cf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b9348cf82b182d6681db12b794dabd7bf80b8fb/pkg/jobs/structured_log.go#L32)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

Added an `EventLevel` field. I wasn't sure what exactly you had in mind in terms of what the different levels would mean, so I went with 1 being the most "severe" and 5 being the least. I also used nomenclature similar to log levels. Lmk if you have different ideas / opinions about this.
</blockquote></details>

We already have log severity functionality, can't we use this here? Where you call `log.InfoE`
___
*[`pkg/util/log/structured_v2.go` line 128 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0stv0R0Braxak_ueDU:-O0stv0R0Braxak_ueDV:bd6l17u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a659f8ca5638ded03aba088ea3a1e3926709a1/pkg/util/log/structured_v2.go#L128)):*
> ```Go
> // structuredEvent logs a logpb.EventPayload as a structured entry at a provided
> // logpb.Severity. Once it is logged, it is sent to be processed by a StructuredLogProcessor.
> func logStructuredEvent(
> ```

This gets a bit confusing since it's so similar to `log.StructuredEvent`. can we combine the two into a single implementation?
___
*[`pkg/util/log/eventpb/job_events.proto` line 59 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0srgHN08qoCdLYB434:-O0srgHN08qoCdLYB435:bb2185u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a659f8ca5638ded03aba088ea3a1e3926709a1/pkg/util/log/eventpb/job_events.proto#L59)):*
> ```proto
> 
>   // A human parsable description of the status change
>   string description = 4 [(gogoproto.jsontag) = ",omitempty", (gogoproto.moretags) = "redact:\"nonsensitive\""];
> ```

I don't think we should assume that the description is non-sensitive since it's pretty freeform. Do we have support for RedactableString types in these protos? that would be ideal.

Use this as a template: https://github.com/cockroachdb/cockroach/blob/master/pkg/util/log/eventpb/sql_audit_events.proto#L34

Then when constructing the value use `redact.Sprintf` and that will properly redact any string params.


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/jobs/structured_log.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1638186023:-O0sxeSn-D0huf4zTHfk:b-ffe4j4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b9348cf82b182d6681db12b794dabd7bf80b8fb/pkg/jobs/structured_log.go#L32)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

We already have log severity functionality, can't we use this here? Where you call `log.InfoE`
</blockquote></details>

@dhartunian are you commenting on having an `EventLevel` field, or are you talking about using the existing log severity instead of the new `log.InfoE` function? 
___
*[`pkg/util/log/structured_v2.go` line 128 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0stv0R0Braxak_ueDU:-O0syC0d2bm8GWeMWeex:b-a0bs6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a659f8ca5638ded03aba088ea3a1e3926709a1/pkg/util/log/structured_v2.go#L128)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This gets a bit confusing since it's so similar to `log.StructuredEvent`. can we combine the two into a single implementation?
</blockquote></details>

I wasn't sure about the best way to go about this. There are were 3 main things that dissuaded me from using the existing `log.StructuredEvent`.

1. The ability to choose the log level. `log.StructuredEvent` always logs at INFO severity, and I think it would be useful to choose the log level of the event.
2. `log.StructuredEvent` calls `makeStructuredEntry` with a depth of 0, so all logs show the file `event_log.go:48`
3. I didn't want to add a call to `logging.processStructured` in `log.StructuredEvent` because this function is already wrapped by `pkg/sql/event_log` to write to the event log table.

Given that `log.StructuredEvent` is only called in 33 places, i think it would be okay to change the signature from:
 `func StructuredEvent(ctx context.Context, event logpb.EventPayload) `
to:
`func StructuredEvent(ctx context.Context, event logpb.EventPayload, sev logpb.Severity, depth int, )`

But I dont really want to add the call `logging.processStructured()` since that would require adding a `bool` argument to flag the logic for existing callers. 

What do you think of making the proposed update to the function signature, and changing the wrapping `InfoE`, `WarnE`, etc. functions to provide the `logging.processStructured()` function call?
___
*[`pkg/util/log/eventpb/job_events.proto` line 59 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0srgHN08qoCdLYB434:-O0sx_dM89JV-b0oDRSL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a659f8ca5638ded03aba088ea3a1e3926709a1/pkg/util/log/eventpb/job_events.proto#L59)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I don't think we should assume that the description is non-sensitive since it's pretty freeform. Do we have support for RedactableString types in these protos? that would be ideal.

Use this as a template: https://github.com/cockroachdb/cockroach/blob/master/pkg/util/log/eventpb/sql_audit_events.proto#L34

Then when constructing the value use `redact.Sprintf` and that will properly redact any string params.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/jobs/structured_log.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125319#gh-1638186023:-O0t44zh-yL_WvCRhO6C:b8rhhr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b9348cf82b182d6681db12b794dabd7bf80b8fb/pkg/jobs/structured_log.go#L32)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

@dhartunian are you commenting on having an `EventLevel` field, or are you talking about using the existing log severity instead of the new `log.InfoE` function?
</blockquote></details>

Just and FYI, this struct no longer exists.


<!-- Sent from Reviewable.io -->

kyle-a-wong committed:
log: add new parameter to log.StructuredEvent

adds a severity parameter to
`log.StructuredEvent` to give more fine
grained control of logged structured events

Epic: none
Release note: None

dhartunian reviewed (commented):
> Reviewed 1 of 1 files at r9, 1 of 3 files at r10, 41 of 41 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @dt, @kyle-a-wong, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/util/log/event_log.go` line 23 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0uiOVqBA7rbOh-pSOI:-O0uiOVqBA7rbOh-pSOJ:bdwkvf6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ec4aac60adb2ff127edfe7b5a4148129fc8f1f2/pkg/util/log/event_log.go#L23)):*
> ```Go
> // StructuredEvent emits a structured event log of severity sev to the channel the provided
> // event belongs to.
> func StructuredEvent(ctx context.Context, event logpb.EventPayload, sev logpb.Severity, depth int) {
> ```

I'd put `sev` as the second arg before `event`
___
*[`pkg/util/log/event_log.go` line 36 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0uiQeP4ikmTn-PEFJw:-O0uiQeP4ikmTn-PEFJx:b-8p2fc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ec4aac60adb2ff127edfe7b5a4148129fc8f1f2/pkg/util/log/event_log.go#L36)):*
> ```Go
> 		sev,
> 		event.LoggingChannel(),
> 		depth,
> ```

I'd reconsider exposing `depth` in this method since typically it's not something an engineer needs to think about when logging. If we really want to was can expose a `StructuredEventDepth` alternative which mimics what we already have (`Infof` and `InfofDepth`)
___
*[`pkg/util/log/eventpb/job_events.proto` line 70 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0uj-LI-8CToT963FdW:-O0uj-LJ--wYzEKfLMeu:b7u9yq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2bad2909dab65dd66e8c99aded0b627b7117ef4/pkg/util/log/eventpb/job_events.proto#L70)):*
> ```proto
>   int32 run_num = 7 [(gogoproto.jsontag) = ",omitempty"];
> 
>   // An error that may have occurred while the job was running.
> ```

nit: I think some more text is needed to disambiguate between the two errors. Does this error mean that the job could continue? retry? Regarding the message below, is intervention by the customer necessary?
___
*[`pkg/jobs/jobs_test.go` line 386 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0uipBz1CCl4HRsCS50:-O0uipBz1CCl4HRsCS51:b-nrqx74) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2bad2909dab65dd66e8c99aded0b627b7117ef4/pkg/jobs/jobs_test.go#L386)):*
> ```Go
> 	// Requires.eventually is needed because status change events occur in a transaction commit trigger
> 	// and not synchronously.
> 	require.Eventually(t, func() bool {
> ```

We use `testutils.SucceedsSoon` instead of `require.Eventually`.

Sidenote: would be really nice not to have to do this and for the test to succeed synchronously. Not sure if that's doable.


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/util/log/event_log.go` line 23 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0uiOVqBA7rbOh-pSOI:-O1Hdi5j5Rm2siwjwKic:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ec4aac60adb2ff127edfe7b5a4148129fc8f1f2/pkg/util/log/event_log.go#L23)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I'd put `sev` as the second arg before `event`
</blockquote></details>

Done.
___
*[`pkg/util/log/event_log.go` line 36 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0uiQeP4ikmTn-PEFJw:-O1Hdh2Y07__9MPaf_B1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ec4aac60adb2ff127edfe7b5a4148129fc8f1f2/pkg/util/log/event_log.go#L36)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I'd reconsider exposing `depth` in this method since typically it's not something an engineer needs to think about when logging. If we really want to was can expose a `StructuredEventDepth` alternative which mimics what we already have (`Infof` and `InfofDepth`)
</blockquote></details>

Done.
___
*[`pkg/jobs/jobs_test.go` line 386 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0uipBz1CCl4HRsCS50:-O1Hdz7p47A2wHkguvUE:b-cojgye) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2bad2909dab65dd66e8c99aded0b627b7117ef4/pkg/jobs/jobs_test.go#L386)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

We use `testutils.SucceedsSoon` instead of `require.Eventually`.

Sidenote: would be really nice not to have to do this and for the test to succeed synchronously. Not sure if that's doable.
</blockquote></details>

Done.

Regarding the side note, I did some testing and its actually not related to the commit trigger, so I removed that comment. Due to the nature of these tests, I think we need to keep `testutils.SucceedsSoon`. `func (rts *registryTestSuite) check(t *testing.T, expectedStatus jobs.Status)` similarly uses `SucceedsSoon` to assert that the database has been updated with the correct status. I'm not sure if there is a way to make it so we can test synchronously, but I can poke around a little more and see if I can make it work


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @msbutler, @stevendanna, and @xinhaoz)

___
*[`pkg/util/log/eventpb/job_events.proto` line 70 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0uj-LI-8CToT963FdW:-O1HrgprEJnynWapPIx9:b-vdktqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2bad2909dab65dd66e8c99aded0b627b7117ef4/pkg/util/log/eventpb/job_events.proto#L70)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: I think some more text is needed to disambiguate between the two errors. Does this error mean that the job could continue? retry? Regarding the message below, is intervention by the customer necessary?
</blockquote></details>

This error field isn't very well documented in the code so it's not clear to me what exactly the error field here may mean. In fact it has the following comment in the proto file:
```
  // TODO (lucy): Deprecate the string error field and move to using the encoded
  // errors everywhere.
```

But as per a comment from michael butler here: https://github.com/cockroachdb/cockroach/pull/124242#discussion_r1611872829 , he suggested we only use these 2 errors. 

Looking for usages, I found that this is set in the payload when a job goes into reverted, failed, revertedFailed, and cancelled status. 


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 11 files at r12, 30 of 30 files at r13, 2 of 11 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-:-O1HwXA22mbgo2M9vROg:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @kyle-a-wong, @msbutler, @stevendanna, and @xinhaoz)

___
*[`-- commits` line 4 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O1HveRCExc4OC4xZTvd):*
nit: "parameter"
___
*[`pkg/util/log/structured_v2.go` line 128 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/125319#-O0stv0R0Braxak_ueDU:-O1HwLC8346BerIvj4t1:b-nk2jd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a659f8ca5638ded03aba088ea3a1e3926709a1/pkg/util/log/structured_v2.go#L128)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

I wasn't sure about the best way to go about this. There are were 3 main things that dissuaded me from using the existing `log.StructuredEvent`.

1. The ability to choose the log level. `log.StructuredEvent` always logs at INFO severity, and I think it would be useful to choose the log level of the event.
2. `log.StructuredEvent` calls `makeStructuredEntry` with a depth of 0, so all logs show the file `event_log.go:48`
3. I didn't want to add a call to `logging.processStructured` in `log.StructuredEvent` because this function is already wrapped by `pkg/sql/event_log` to write to the event log table.

Given that `log.StructuredEvent` is only called in 33 places, i think it would be okay to change the signature from:
 `func StructuredEvent(ctx context.Context, event logpb.EventPayload) `
to:
`func StructuredEvent(ctx context.Context, event logpb.EventPayload, sev logpb.Severity, depth int, )`

But I dont really want to add the call `logging.processStructured()` since that would require adding a `bool` argument to flag the logic for existing callers. 

What do you think of making the proposed update to the function signature, and changing the wrapping `InfoE`, `WarnE`, etc. functions to provide the `logging.processStructured()` function call?
</blockquote></details>

We went ahead with changing function signature. Thx!


<!-- Sent from Reviewable.io -->
